File: extension/src/api/workspaceEdit.ts
import { agentHttpClient } from "./client";

export type WorkspaceFileInput = {
    file_path: string;
    original_text: string | null;
};

export type WorkspaceEditRequest = {
    instruction: string;
    files: WorkspaceFileInput[];
    user_context?: string | null;
    max_files?: number;
};

export type WorkspaceFileOutput = {
    file_path: string;
    action: "create" | "modify";
    updated_text: string;
    unified_diff: string;
    warnings: string[];
};

export type WorkspaceEditResponse = {
    operations: WorkspaceFileOutput[];
    warnings: string[];
};

export async function workspaceEdit(req: WorkspaceEditRequest): Promise<WorkspaceEditResponse> {
    const client = agentHttpClient();
    const resp = await client.post("/workspace_edit", req);
    return resp.data as WorkspaceEditResponse;
}

File: agent/app/models/workspace_edit.py
from pydantic import BaseModel, Field
from typing import Optional, Literal, List

class WorkspaceFileInput(BaseModel):
    file_path: str = Field(..., min_length=1)
    original_text: Optional[str] = None
    
class WorkspaceEditRequest(BaseModel):
    instruction: str = Field(..., min_length=1)
    files: List[WorkspaceFileInput] = Field(..., min_length=1)
    user_context: Optional[str] = None
    max_files: int = 10
    
class WorkspaceFileOutput(BaseModel):
    file_path: str
    action: Literal["create", "modify"]
    updated_text: str
    unified_diff: str = ""
    warnings: List[str] = []

class WorkspaceEditResponse(BaseModel):
    operations: List[WorkspaceFileOutput]
    warnings: List[str] = []
    

    